examples/step-37/step-37.cc

```diff
@@ -774,283 +874,247 @@ namespace Step37
 
-    solution.reinit (dof_handler.n_dofs());
-    system_rhs.reinit (dof_handler.n_dofs());
+    system_matrix.initialize_dof_vector(solution);
+    system_matrix.initialize_dof_vector(system_rhs);
 
     setup_time += time.wall_time();
     time_details << "Setup matrix-free system   (CPU/wall) "
                  << time() << "s/" << time.wall_time() << "s" << std::endl;
     time.restart();
 
     // Next, initialize the matrices for the multigrid method on all the
-    // levels. The function MGTools::make_boundary_list returns for each
-    // multigrid level which degrees of freedom are located on a Dirichlet
-    // boundary; we force these DoFs to have value zero by adding to the
-    // ConstraintMatrix object a zero condition by using the command
-    // ConstraintMatrix::add_line. Once this is done, we close the
-    // ConstraintMatrix on each level so it can be used to read out indices
-    // internally in the MatrixFree.
-    const unsigned int nlevels = triangulation.n_levels();
+    // levels. The data structure MGConstrainedDoFs keeps information about
+    // the indices subject to boundary conditions as well as the indices on
+    // edges between different refinement levels as described in the step-16
+    // tutorial program. We then go through the levels of the mesh and
+    // construct the constraints and matrices on each level. These follow
+    // closely the construction of the system matrix on the original mesh,
+    // except the slight difference in naming when accessing information on
+    // the levels rather than the active cells.
+    const unsigned int nlevels = triangulation.n_global_levels();
     mg_matrices.resize(0, nlevels-1);
+    mg_mf_storage.resize(0, nlevels-1);
     mg_constraints.resize (0, nlevels-1);
 
-    typename FunctionMap<dim>::type dirichlet_boundary;
-    ZeroFunction<dim>               homogeneous_dirichlet_bc (1);
-    dirichlet_boundary[0] = &homogeneous_dirichlet_bc;
-    std::vector<std::set<types::global_dof_index> > boundary_indices(triangulation.n_levels());
-    MGTools::make_boundary_list (dof_handler,
-                                 dirichlet_boundary,
-                                 boundary_indices);
+    std::set<types::boundary_id> dirichlet_boundary;
+    dirichlet_boundary.insert(0);
+    mg_constrained_dofs.initialize(dof_handler);
+    mg_constrained_dofs.make_zero_boundary_constraints(dof_handler, dirichlet_boundary);
+
     for (unsigned int level=0; level<nlevels; ++level)
       {
-        std::set<types::global_dof_index>::iterator bc_it = boundary_indices[level].begin();
-        for ( ; bc_it != boundary_indices[level].end(); ++bc_it)
-          mg_constraints[level].add_line(*bc_it);
-
+        IndexSet relevant_dofs;
+        DoFTools::extract_locally_relevant_level_dofs(dof_handler, level,
+                                                      relevant_dofs);
+        mg_constraints[level].reinit(relevant_dofs);
```

Comment:
do you actually need to keep around `mg_constraints` for each level? I think you can do 
```
ConstraintMatrix level_constraints;
```
within this loop and initialize the data accordingly. So no need for a member variable.